### PR TITLE
fix caller and callee both simulcast or just one side simucast fail and fix the substream status show error on the webpage

### PR DESCRIPTION
**Pull request by lucylu-star**
1. fix caller and callee both simulcast fail.
2. support just callee  simulcast ok.
3. fix the substream status show error on the webpage.

### CODE REVIEW

**Comment by lminiero** [comment]
Simulcast cannot be enabled by the receiving party, as far as I know, because the simulcast envelope is created when the transceiver is created. Since the transceiver is created by the PeerConnection and not by the browser, in case of a callee, simulcast cannot be added there.

**Comment by lucylu-star** [comment]
As you said, callee cannot simulcast.  I revert some.
But, now have the bug that is caller simulcast failed and substream level status shows error when videocall.
I think the new pull request fix the issue.

**Comment by lucylu-star** [comment]
Only modify  janus_videocall.c

File: plugins/janus_videocall.c [hunk 1]
```diff
@@ -786,7 +786,7 @@ void janus_videocall_incoming_rtp(janus_plugin_session *handle, janus_plugin_rtp
 				json_t *result = json_object();
 				json_object_set_new(result, "event", json_string("simulcast"));
 				json_object_set_new(result, "videocodec", json_string(janus_videocodec_name(session->vcodec)));
-				json_object_set_new(result, "substream", json_integer(session->sim_context.substream));
+				json_object_set_new(result, "substream", json_integer(peer->sim_context.substream));
```
Comment (lminiero) [inline plugins/janus_videocall.c:789]: This is wrong. The substream returned here is the substream the user (so `session`) wants from the peer. Substream and temporal info are always receiver related.

File: plugins/janus_videocall.c [hunk 2]
```diff
@@ -798,7 +798,7 @@ void janus_videocall_incoming_rtp(janus_plugin_session *handle, janus_plugin_rtp
 				json_t *result = json_object();
 				json_object_set_new(result, "event", json_string("simulcast"));
 				json_object_set_new(result, "videocodec", json_string(janus_videocodec_name(session->vcodec)));
-				json_object_set_new(result, "temporal", json_integer(session->sim_context.templayer));
+				json_object_set_new(result, "temporal", json_integer(peer->sim_context.templayer));
```
Comment (lminiero) [inline plugins/janus_videocall.c:801]: Same here.

File: plugins/janus_videocall.c [hunk 3]
```diff
@@ -1358,20 +1358,15 @@ static void *janus_videocall_handler(void *data) {
 			json_t *msg_simulcast = json_object_get(msg->jsep, "simulcast");
 			if(msg_simulcast && janus_get_codec_pt(msg_sdp, "vp8") > 0) {
 				JANUS_LOG(LOG_VERB, "VideoCall callee (%s) is going to do simulcasting\n", session->username);
-				session->ssrc[0] = json_integer_value(json_object_get(msg_simulcast, "ssrc-0"));
-				session->ssrc[1] = json_integer_value(json_object_get(msg_simulcast, "ssrc-1"));
-				session->ssrc[2] = json_integer_value(json_object_get(msg_simulcast, "ssrc-2"));
+		        int rid_ext_id = -1;
+				janus_rtp_simulcasting_prepare(msg_simulcast, &rid_ext_id, session->ssrc, session->rid);
+				session->sim_context.rid_ext_id = rid_ext_id;
```
Comment (lminiero) [inline plugins/janus_videocall.c:1363]: As I pointed out, callees cannot do simulcast. It's likely the whole code above (the one already in the repo) is redundant and should be removed.

File: plugins/janus_videocall.c [hunk 4]
```diff
@@ -1358,20 +1358,15 @@ static void *janus_videocall_handler(void *data) {
...
 				int i=0;
 				for(i=0; i<3; i++) {
 					session->ssrc[i] = 0;
 					g_free(session->rid[0]);
 					session->rid[0] = NULL;
-					if(peer) {
-						peer->ssrc[i] = 0;
-						g_free(peer->rid[0]);
-						peer->rid[0] = NULL;
-					}
```
Comment (lminiero) [inline plugins/janus_videocall.c:1374]: This is the only fix that I think does belong here: in fact, it was incorrectly disabling the previously configured simulcast settings for the peer (the caller in this instance).

**Review by lminiero — COMMENTED** [review]
Added some new notes inline. Thanks!

File: plugins/janus_videocall.c (hunk 1 above)
Comment (lucylu-star) [reply plugins/janus_videocall.c:789]: disagree. 
 I think peer->sim_context.substream is right.   you can check it seeing the substream status on videocall webpage.  I have verified it.   Using peer, the showing is ok.  if using session, the status showing is abnormal.
Peer is the callee.   The substream returned here is what substream the callee wants. It is showed on webpage.

File: plugins/janus_videocall.c (hunk 3 above)
Comment (lucylu-star) [reply plugins/janus_videocall.c:1363]: yes, I will remove it.

File: plugins/janus_videocall.c (hunk 4 above)
Comment (lucylu-star) [reply plugins/janus_videocall.c:1374]: yes

File: plugins/janus_videocall.c (hunk 1 above)
Comment (lminiero) [reply plugins/janus_videocall.c:789]: Ah you're rght, sorry: this is an event we're sending to `peer`, not `session` as I thought :+1:   The `session->vcodec` above threw me off.

File: plugins/janus_videocall.c [hunk 5]
```diff
@@ -1358,20 +1358,15 @@ static void *janus_videocall_handler(void *data) {
 			json_t *msg_simulcast = json_object_get(msg->jsep, "simulcast");
 			if(msg_simulcast && janus_get_codec_pt(msg_sdp, "vp8") > 0) {
 				JANUS_LOG(LOG_VERB, "VideoCall callee (%s) is going to do simulcasting\n", session->username);
-				session->ssrc[0] = json_integer_value(json_object_get(msg_simulcast, "ssrc-0"));
-				session->ssrc[1] = json_integer_value(json_object_get(msg_simulcast, "ssrc-1"));
-				session->ssrc[2] = json_integer_value(json_object_get(msg_simulcast, "ssrc-2"));
+		        int rid_ext_id = -1;
```
Comment (lminiero) [inline plugins/janus_videocall.c:1361]: Indentation is broken here: it should be tabs, not spaces.

File: plugins/janus_videocall.c (hunk 3 above)
Comment (lminiero) [reply plugins/janus_videocall.c:1363]: As soon as you remove tis part here, it's good to merge for me :+1:

Comment (lminiero) [reply plugins/janus_videocall.c:1363]: Note: I can get rid of the whole existing (and unneeded) simulcast check for callees myself later.

**Comment by lucylu-star** [comment]
just remove several unneeded lines. Why auto check failed?

**Comment by lminiero** [comment]
It looks like an issue pulling libnice from gitlab, so not caused by your patch. Hopefully it was a temporary error, I'll try to restart the automated build later.

**Comment by lminiero** [comment]
Thanks, merging! :+1:

**Pull request merged**